### PR TITLE
[FEATURE] Eliminate the use of Validator in creating Profiler, DomainBuilder, and ParameterBuilder classes of the Rule Based Profiler

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Scope: Significant Simplification of the `Rule Based Profiler` architecture
Practical and elegant symmetry, while maintaining the appropriate separation of concerns, between `DomainBuilder` and `ParameterBuilder` functionalities has been achieved.

* The `Validator` object from the user's pipeline/notebook no longer needs to be passed to the `Profiler` constructor.
* Both `DomainBuilder` and `ParameterBuilder` configuration can define the appropriate `batch_request` configuration as needed for their specific purposes -- to build the domain and to calculate the parameter values, respectively.
* As with other configuration entities, `batch_request` can be associated with a variable to be reused throughout the YAML configuration (helps creating more clear and concise YAML configuration files).
* Eliminated the `MultibatchParameterBuilder` (it is not used -- inheriting from the top-level `ParameterBuilder` is simpler).
* Refactoring more common functionality into the utility module.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
-
-
-

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/rule_based_profiler/domain_builder/column_domain_builder.py [hunk 1]
```diff
@@ -1,35 +1,33 @@
...
 from great_expectations.validator.validation_graph import MetricConfiguration
-from great_expectations.validator.validator import Validator
 
 
 class ColumnDomainBuilder(DomainBuilder):
     def _get_domains(
         self,
-        *,
-        validator: Optional[Validator] = None,
+        variables: Optional[ParameterContainer] = None,
```
Comment (talagluck) [inline great_expectations/rule_based_profiler/domain_builder/column_domain_builder.py:17]: Is the addition of `variables` in the `_get_domains` methods to insure that variables are accessible even before we have `ParameterBuilders`?

File: great_expectations/rule_based_profiler/domain_builder/domain.py [hunk 2]
```diff
@@ -80,7 +80,11 @@ def __eq__(self, other):
                 hasattr(other, "to_json_dict")
                 and self.to_json_dict() == other.to_json_dict()
             )
-            or (isinstance(other, dict) and self.to_json_dict() == other)
+            or (
+                isinstance(other, dict)
+                and self.to_json_dict()
+                == filter_properties_dict(properties=other, clean_falsy=True)
```
Comment (talagluck) [inline great_expectations/rule_based_profiler/domain_builder/domain.py:86]: What is the reason for this change?

File: great_expectations/rule_based_profiler/domain_builder/domain_builder.py [hunk 3]
```diff
@@ -1,53 +1,111 @@
...
 
-        IMPORTANT: If an implementation of DomainBuilder sets "batch_id": my_batch_id in "Domain.domain_kwargs" and also
-        calls "validator.get_metric()" in its logic, then "MetricConfiguration" must also set "batch_id" as follows:
-        validator.get_metric(
-            metric=MetricConfiguration(
-                metric_name="my_metric",
-                metric_domain_kwargs={key_mdkw0: value_mdkw0, key_mdkw1: value_mdkw1, "batch_id": my_batch_id,},
-                metric_value_kwargs={key_mvkw0: value_mvkw0, key_mvkw1: value_mvkw1,},
-                metric_dependencies={key_mdeps0: value_mdeps0, key_mdeps1: value_mdeps1,},
+    def get_batch_id(
```
Comment (talagluck) [inline great_expectations/rule_based_profiler/domain_builder/domain_builder.py:61]: Given that this method will raise an error if there is more than one batch in the `BatchRequest`, I wonder if we should move it to the child classes. I can imagine a future scenario where we have a MultiBatchDomainBuilder, and so we wouldn't want to raise an error in this case. Thoughts?

File: great_expectations/rule_based_profiler/domain_builder/simple_column_suffix_domain_builder.py [hunk 4]
```diff
@@ -68,8 +85,10 @@ def _get_domains(
             Domain(
                 domain_type=MetricDomainTypes.COLUMN,
                 domain_kwargs={
-                    "column": column_name,
-                    "batch_id": validator.active_batch_id,
+                    **{
+                        "column": column_name,
+                    },
```
Comment (talagluck) [inline great_expectations/rule_based_profiler/domain_builder/simple_column_suffix_domain_builder.py]: Why use this pattern instead of `"column": column_name`?

File: great_expectations/rule_based_profiler/domain_builder/simple_column_suffix_domain_builder.py [hunk 5]
```diff
@@ -43,12 +57,15 @@ def _get_domains(
                     "Unrecognized column_name_suffixes directive -- must be a list or a string."
                 )
 
-        table_column_names: List[str] = validator.get_metric(
+        metric_domain_kwargs: dict = {
+            "batch_id": self.get_batch_id(variables=variables),
+        }
```
Comment (talagluck) [inline great_expectations/rule_based_profiler/domain_builder/simple_column_suffix_domain_builder.py:65]: This is maybe a philosophical question, and maybe out of scope: Can we always assume that a Batch is tabular? Currently, a `batch_id` is sufficient as the only key passed into `metric_domain_kwargs` because we know that a `batch_id` represents a clear single table. But if in a future world a Batch might be able to take other shapes, span multiple tables, etc. I wonder if we need some sort of method to validate that a Batch is a table so that we know we can rely on this here.

File: great_expectations/rule_based_profiler/parameter_builder/parameter_container.py [hunk 6]
```diff
@@ -251,7 +251,7 @@ def _build_parameter_node_tree_for_one_parameter(
 
 def get_parameter_value_by_fully_qualified_parameter_name(
     fully_qualified_parameter_name: str,
-    domain: Domain,
+    domain: Optional[Domain] = None,
```
Comment (talagluck) [inline great_expectations/rule_based_profiler/parameter_builder/parameter_container.py:254]: Why is this optional now?

File: tests/rule_based_profiler/bobby_user_workflow_fixture.py [hunk 7]
```diff
@@ -22,26 +22,6 @@ def bobby_columnar_table_multi_batch():
...
-                        "metric_configuration": {
-                            "metric_name": "table.row_count",
-                            "metric_domain_kwargs": {
-                                "batch_id": "021563e94d7866f395288f6e306aed9b"
-                            },
-                        }
-                    }
-                },
-            }
-        ),
```
Comment (talagluck) [inline tests/rule_based_profiler/bobby_user_workflow_fixture.py:44]: Why was this removed?

**Review by talagluck — COMMENTED** [review]
This is lovely! Love the consistency that this brings around BatchRequests - the code feels much more streamlined and logical. Just a few questions.

File: great_expectations/rule_based_profiler/domain_builder/column_domain_builder.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/domain_builder/column_domain_builder.py:17]: @talagluck Yes!  And once we move `variables` outside and above the `Rule`, then perhaps we might be able to simplify this even further (I would recommend this to be done as part of a separate PR).  Thanks.

File: great_expectations/rule_based_profiler/domain_builder/domain.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/domain_builder/domain.py:86]: It fixes a bug that was not manifested before -- comparison of `Domain` objects missed a case.

File: great_expectations/rule_based_profiler/domain_builder/domain_builder.py (hunk 3 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/domain_builder/domain_builder.py:61]: @talagluck In general, I do not disagree.  Moreover, I made a small but useful refactor of this class as part of another PR (will come after this one is merged).  How about we reassess at that time?  It should be relatively straightforward.  Thanks.

File: great_expectations/rule_based_profiler/domain_builder/simple_column_suffix_domain_builder.py (hunk 4 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/domain_builder/simple_column_suffix_domain_builder.py]: @talagluck Good question -- this pattern improves code robustness.  We define the common components of the dictionary up front and then use to instantiate the final dictionary as a composition of input dictionaries (one of which contains the common component)  -- and this is the correct syntax (as supported by Python).  Thanks.

File: great_expectations/rule_based_profiler/domain_builder/simple_column_suffix_domain_builder.py (hunk 5 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/domain_builder/simple_column_suffix_domain_builder.py:65]: @talagluck This is a good question -- the notion of what constitutes a batch has been going through constant stream of questions.  Because they have not been answered definitively yet, I believe that we can leave the implementation as is for now.  However, here are some of the questions that should be addressed in the coming months/quarters:
* Are we only considering tabular data?  If not, what other kind of data should we figure out how to represent?
* When we partition a large physical table, do we end up with one batch consisting of multiple partitions?  Or is each partition actually a batch?  (This is a very important question, because it focuses the discussion on what is being validated.  In other words, the user wants to validate a table, which happens to be large; hence, the `Validator` should represent that large table; however, if a batch represents a smaller partition of that table and the `Validator` represents that batch, then how do we help the user validate the actual large table of import to the user?)
* Why did we remove `batch_id` from the `TableDomainBuilder`?  If the user can now specify the `batch_request` in the domain builder's configuration, then why would we take away the ability of the users to build the domain out of any batch that interests them?  Since the `Validator` will accept an explicit `batch_id`, then the `TableDomainBuilder` should support the ability to set that `batch_id` from the `Batch` returned by processing the configured batch_request.  (I propose we add this back in a small follow on PR.)

File: great_expectations/rule_based_profiler/parameter_builder/parameter_container.py (hunk 6 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/parameter_builder/parameter_container.py:254]: @talagluck This is required for the very reason you mentioned before: the method `get_parameter_value_by_fully_qualified_parameter_name()` should be useful to retrieve values from `variables` -- we use this in the `DomainBuilder` classes, where, by definition, the `Domain` does not yet exist (it is being built).  I suspect that at some point we will gather a sufficient amount of justification to introduce a proper `RuleExecutionState` class, which will handle a number of these cases inside (I feel that we can defer it for now).

File: tests/rule_based_profiler/bobby_user_workflow_fixture.py (hunk 7 above)
Comment (alexsherstinsky) [reply tests/rule_based_profiler/bobby_user_workflow_fixture.py:44]: @talagluck It was duplicated code.  The cleaned up version here clearly separates the 1 table expectation from the 30 column expectations.

File: great_expectations/rule_based_profiler/domain_builder/table_domain_builder.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    ParameterContainer,
+)
+
+
+class TableDomainBuilder(DomainBuilder):
+    """
+    The interface method of TableDomainBuilder emits a single Domain object, corresponding to the implied Batch (table).
+
+    Note that for appropriate use-cases, it should be readily possible to build a multi-batch implementation, where a
+    separate Domain object is emitted for each individual Batch (using its respective batch_id).  (This is future work.)
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/domain_builder/table_domain_builder.py:18]: Since this is not yet a viable feature, is there any value in omitting it from the docstring for now (or moving it to a TODO)?

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py [hunk 9]
```diff
@@ -145,25 +144,20 @@ def _build_parameters(
...
-        batch_ids_for_metrics_calculations: Optional[
-            List[str]
-        ] = self.get_batch_ids_for_metrics_calculations(
+        batch_ids: Optional[List[str]] = self.get_batch_ids(
             domain=domain,
             variables=variables,
             parameters=parameters,
         )
-        if not batch_ids_for_metrics_calculations:
+        if not batch_ids:
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:152]: This is more for my curiosity than a review note but what's the difference between `if not x` versus `if x is None`? I've seen both used in the GE codebase but is there value in the explicit `if x is None` as opposed to relying on Python's concept of truthy/falsy?

**Review by cdkini — COMMENTED** [review]
This looks like a very thorough refactor, awesome work! Primarily reviewed your discussions with @talagluck but added a few stylistic comments. 

Still wrapping my head around some concepts in the profiler but I'd love to touch on some of the open questions noted in the comments and see how they might impact our team's efforts these next few months.

File: great_expectations/rule_based_profiler/domain_builder/domain_builder.py (hunk 3 above)
Comment (talagluck) [reply great_expectations/rule_based_profiler/domain_builder/domain_builder.py:61]: Sounds reasonable!

File: great_expectations/rule_based_profiler/domain_builder/simple_column_suffix_domain_builder.py (hunk 4 above)
Comment (talagluck) [reply great_expectations/rule_based_profiler/domain_builder/simple_column_suffix_domain_builder.py]: I'd love to chat more about this. I can understand unpacking `metric_domain_kwargs` and passing it in, but creating a dictionary to immediately unpack it here feels confusing to me.

File: great_expectations/rule_based_profiler/domain_builder/simple_column_suffix_domain_builder.py (hunk 5 above)
Comment (talagluck) [reply great_expectations/rule_based_profiler/domain_builder/simple_column_suffix_domain_builder.py:65]: These are great questions!

To the last one, please check my understanding: `TableDomainBuilder` feels like a pretty specific case. As a reminder, `DomainBuilders` determine the DomainKwargs to use for building an ExpectationConfiguration. But table-level expectations don't need to take DomainKwargs - since Validators can currently only validate one Batch at a time, we just assume that a table-level expectation applies to the entire Batch. So `TableDomainBuilder` winds up being almost like a dummy Domain builder, saying the Domain for this ExpectationConfiguration is the entire Batch. From my understanding, creating an ExpectationConfiguration where a `batch_id` is passed in as a `DomainKwarg` will make it so that that expectation will only be validated when run on a Batch with that `batch_id` (which seems like an unusual case that most people wouldn't want). So I think to avoid that, we should keep this as is, and if we think that there is a use case for building a TableDomain for a specific `batch_id`, we should add a new DomainBuilder.

File: great_expectations/rule_based_profiler/domain_builder/table_domain_builder.py (hunk 8 above)
Comment (talagluck) [reply great_expectations/rule_based_profiler/domain_builder/table_domain_builder.py:18]: Agreed! I think this might be good for the Jira backlog

File: great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py (hunk 9 above)
Comment (talagluck) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:152]: You may have already stated this, but my understanding is that `if not x` will not trigger if `x` is any falsy value (e.g. an empty list or dictionary), where `if x is None` would not trigger in that case (only triggering if `x` is an actual `NoneType`. So it depends on the value of variables that we're dealing with in a particular instance.

Comment (cdkini) [reply great_expectations/rule_based_profiler/parameter_builder/numeric_metric_range_multi_batch_parameter_builder.py:152]: That makes sense! I guess my original thought that they were entirely interchangeable is not necessarily true.

**Review by talagluck — APPROVED** [review]
LGTM - thanks for the changes!

**Review by cdkini — APPROVED** [review]
LGTM as well.

Should we note this discussion on the treatment of `batch_id` in Jira or just reconvene on that at another time?
